astor/code_gen.py

```diff
@@ -530,15 +531,61 @@ def write_comma():
     def visit_Name(self, node):
         self.write(node.id)
 
+    # ast.Constant is new in Python 3.6 and it replaces ast.Bytes,
+    # ast.Ellipsis, ast.NameConstant, ast.Num, ast.Str in Python 3.8
+    def visit_Constant(self, node):
+        value = node.value
+
+        if isinstance(value, (float, complex)):
+            self._handle_numeric_constant(value)
+        elif isinstance(value, str):
+            precedence = self.get__pp(node)
+            self._handle_string_constant(value, precedence)
+        elif value is Ellipsis:
+            self.write('...')
+        else:
+            self.write(repr(value))
+
     def visit_JoinedStr(self, node):
-        self.visit_Str(node, True)
+        precedence = self.get__pp(node)
+
+        has_ast_constant = sys.version_info >= (3, 6)
+
+        def recurse(node):
+            for value in node.values:
+                if isinstance(value, ast.Str):
+                    self.write(value.s)
+                elif isinstance(value, ast.FormattedValue):
+                    with self.delimit('{}'):
+                        self.visit(value.value)
+                        if value.conversion != -1:
+                            self.write('!%s' % chr(value.conversion))
+                        if value.format_spec is not None:
+                            self.write(':')
+                            recurse(value.format_spec)
+                elif has_ast_constant and isinstance(value, ast.Constant):
+                    self.write(value.value)
+                else:
+                    kind = type(value).__name__
+                    assert False, 'Invalid node %s inside JoinedStr' % kind
 
-    def visit_Str(self, node, is_joined=False):
+        self._handle_string_constant(node, precedence, join_fn=lambda: recurse(node))
 
+    def _handle_string_constant(self, value, precedence, join_fn=None):
```

Comment:
Why not pass the original `node`? It could be used for precedence and for recursion. You will don't need to repeat `precedence = self.get__pp(node)`.